### PR TITLE
Add Resource "ShootExtensionStatus" in  core.gardener.cloud

### PR DESCRIPTION
**Pull request by danielfoehrKn**
_**How_ to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:

Adds a new resource in the `core.gardener.cloud` resource group with the name `ShootExtensionStatus`.
This resource is intended to hold the information of the `ProviderStatus` field from multiple Gardener Extension resources.

It is modeled after the `ShootState` resource. Hence, there is only a `Spec` (not Status), as no spec is actually reconciled.

**Which issue(s) this PR fixes**:
Part of #3873

**Special notes for your reviewer**:

This PR only contains the type definitions & client generations.

**Release note**:

```feature user
Add the resource `ShootExtensionStatus` to the resource group `core.gardener.cloud` which is intended to hold the information of the `ProviderStatus` field from multiple Gardener Extension resources from the Seed cluster. 
```

### CODE REVIEW

File: pkg/apis/core/v1beta1/types_shootproviderstate.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// of a Gardener extension resource from a Shoot's namespace in the Seed cluster.
+type ExtensionProviderState struct {
+	// Kind of the extension resource
+	Kind string `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// Type of the extension resource
+	// +optional
+	Type *string `json:"type" protobuf:"bytes,2,opt,name=type"`
+	// Status is the provider status of the extension resource
+	// +optional
+	Status *runtime.RawExtension `protobuf:"bytes,3,opt,name=status"`
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/types_shootproviderstate.go]: Why is this optional / a pointer?

File: pkg/apis/core/types_shootproviderstate.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+// ExtensionProviderState contains the kind, the type and the last observed provider status
+// of a Gardener extension resource from a Shoot's namespace in the Seed cluster.
+type ExtensionProviderState struct {
+	// Kind of the extension resource
+	Kind string
+	// Type of the extension resource
+	Type *string
+	// Status is the provider status of the extension resource
+	Status *runtime.RawExtension
```
Comment (timuthy) [inline pkg/apis/core/types_shootproviderstate.go]:
```suggestion
	Status runtime.Object
```

Let's follow the recommendation from the doc string of `RawExtension` (see [here](https://github.com/kubernetes/apimachinery/blob/d23dd757f6df1aa5f84b372fefbe4249d3e59943/pkg/runtime/types.go#L49)).

**Review by timuthy — CHANGES_REQUESTED** [review]
Do you also want to include necessary storage changes for `pkg/registry/core`?

**Comment by rfranzke** [comment]
What about calling the resource `ShootExtensionStatus` instead? Two rationales, first is to not "restrict" it to provider extensions (there might be other extensions like networking or what not that might produce some status), second is to not confuse it with the "state" which is used for control plane migration (also, in the `extensions.gardener.cloud` APIs the fields are called `.status.state` vs. `status.providerStatus`).

File: pkg/apis/core/types_shootproviderstate.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	metav1.ListMeta
+	// Items is the list of ShootProviderStates.
+	Items []ShootProviderState
+}
+
+// ShootProviderStateSpec is the specification of the ShootProviderState.
+type ShootProviderStateSpec struct {
+	// ProviderStatus holds a list containing the information of the providerStatus field of Extension resources
+	// in the Seed cluster
+	ProviderStatus []ExtensionProviderState
```
Comment (rfranzke) [inline pkg/apis/core/types_shootproviderstate.go]:
```suggestion
	ExtensionStatus []ExtensionStatus
```

File: pkg/apis/core/types_shootproviderstate.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// ExtensionProviderState contains the kind, the type and the last observed provider status
+// of a Gardener extension resource from a Shoot's namespace in the Seed cluster.
+type ExtensionProviderState struct {
```
Comment (rfranzke) [inline pkg/apis/core/types_shootproviderstate.go]:
```suggestion
type ExtensionStatus struct {
```

**Comment by danielfoehrKn** [comment]
I added validation and storage implementation in a second commit.

I am a bit torn between naming the plural of the resource "shootextensionstatus" or "shootextensionstatus" - see [here](https://github.com/gardener/gardener/blob/540e1cefbaddec1172fc043c5b2d9a1d67dc47c3/pkg/registry/core/shootextensionstatus/storage/storage.go#L51). On the one hand, `statuses` seems to be the right plural for `status`.
Suggestions welcome.

EDIT: plural being used is "shootextensionstatuses"

File: pkg/apis/core/v1beta1/types_shootproviderstate.go (hunk 1 above)
Comment (danielfoehrKn) [reply pkg/apis/core/v1beta1/types_shootproviderstate.go]: It was modeled after the `ShootState` resource. But I think you are right. If there is no Status, we can remove the entry from the list.

**Comment by danielfoehrKn** [comment]
Looping in @plkokanov to have a look at the patch strategy for the `ExtensionStatus` list as this should be similar like in the `ShootState` and I remember you tweaking something there in the past. Thanks!

**Comment by danielfoehrKn** [comment]
I played around with the strategic-merge-patch on the `extensionStatus` list. 
I encountered problems due to the `rawExtension` field similar to:
`unable to find api field in struct RawExtension for the json field “ec2”`. 

I want to go forward with a `json-merge-patch` with optimistic-locking like it is already implemented for the ShootState resource [in the federated-seed-controller](https://github.com/gardener/gardener/blob/3d5f3f3fd3d65b57f775cb14fb251463ace8b731/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go#L128).
I removed the strategic-merge specific API comments for the list `ExtensionStatus` .

@rfranzke @timuthy

File: pkg/apis/core/types_shootextensionstatus.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	// Standard list object metadata.
+	metav1.ListMeta
+	// Items is a list of ShootExtensionStatus.
+	Items []ShootExtensionStatus
+}
+
+// ShootExtensionStatusSpec is the specification of the ShootExtensionStatus.
+type ShootExtensionStatusSpec struct {
+	// ExtensionStatus holds a list of statuses of Extension resources
+	// in the Seed cluster
```
Comment (rfranzke) [inline pkg/apis/core/types_shootextensionstatus.go]:
```suggestion
	// ExtensionStatus holds a list of statuses of extension controllers.
```

(same for other occurrences)

File: pkg/apis/core/types_shootextensionstatus.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+// ExtensionStatus contains the kind, the type and the last observed status
+// of a Gardener extension resource from a Shoot's namespace in the Seed cluster.
+type ExtensionStatus struct {
+	// Kind of the extension resource
+	Kind string
+	// Type of the extension resource
+	Type string
+	// Status contains the status of the extension resource
+	Status runtime.Object
+}
```
Comment (rfranzke) [inline pkg/apis/core/types_shootextensionstatus.go]: You might also need an optional `Purpose` field

File: pkg/apis/core/v1beta1/register.go [hunk 7]
```diff
@@ -69,6 +69,8 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 		&SeedList{},
 		&Shoot{},
 		&ShootList{},
+		&ShootExtensionStatus{},
+		&ShootExtensionStatusList{},
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/register.go]: What about waiting with `v1beta1` promotion for a while and only add it to `v1alpha1` in the beginning (like done for the `ShootState`)?

File: pkg/apis/core/v1beta1/types_shootextensionstatus.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+// ShootExtensionStatusSpec is the specification of the ShootExtensionStatus.
+type ShootExtensionStatusSpec struct {
+	// ExtensionStatus holds a list of statuses of Extension resources
+	// in the Seed cluster
+	// No strategic merge patch on this list as it contains a rawExtension field
+	// which causes problems during the patch computation
+	// Clients should use a json-merge-patch instead handing over the up-to-date list
+	// +optional
+	ExtensionStatus []ExtensionStatus `json:"extensionStatus,omitempty" protobuf:"bytes,1,rep,name=extensionStatus"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shootextensionstatus.go]:
```suggestion
	Statuses []ExtensionStatus `json:"statuses,omitempty" protobuf:"bytes,1,rep,name=statuses"`
```

(same elsewhere)

File: pkg/apis/core/validation/shootextensionstatus.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+	return allErrs
+}
+
+// ValidateShootExtensionStatusSpecUpdate validates the update to the specification of a ShootExtensionStatus
+func ValidateShootExtensionStatusSpecUpdate(_, _ *core.ShootExtensionStatusSpec) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	return allErrs
+}
```
Comment (rfranzke) [inline pkg/apis/core/validation/shootextensionstatus.go]: Does it make sense to prevent kind/type from being changed?

**Review by rfranzke — COMMENTED** [review]
Can you also add an example YAML file to the `example` folder?

File: pkg/apis/core/v1beta1/register.go (hunk 7 above)
Comment (danielfoehrKn) [reply pkg/apis/core/v1beta1/register.go]: I thought the goal is to phase out the v1alpha1 and not add any more resources to it? Did I misunderstand that?

File: pkg/apis/core/validation/shootextensionstatus.go (hunk 9 above)
Comment (danielfoehrKn) [reply pkg/apis/core/validation/shootextensionstatus.go]: Type possibly yes, but not kind. Kind is the key in the list, so changing it is like adding a new entry and removing the old one.

File: pkg/apis/core/v1beta1/register.go (hunk 7 above)
Comment (danielfoehrKn) [reply pkg/apis/core/v1beta1/register.go]: Anyways, I am fine with changing it to `v1alpha1` if that is the process

**Comment by danielfoehrKn** [comment]
> Can you also add an example YAML file to the example folder?

Shouldn't we maintain only resources in the `example` folder that shall be manually applied / are necessary to set up Gardener?
For instance, the `ShootState` resource is also not in the example folder.

**Review by rfranzke — DISMISSED** [review]
/lgtm
/hold until v1.21 was released

**Review by rfranzke — DISMISSED** [review]

**Comment by danielfoehrKn** [comment]
@timuthy is there something missing from your side?

File: pkg/apis/core/types_shootextensionstatus.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ShootExtensionStatus holds the status information of extensions of a Shoot cluster
+type ShootExtensionStatus struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	// Designed to have an owner reference to the associated Shoot resource
+	metav1.ObjectMeta
+	// Specification of the ShootExtensionStatus.
+	Spec ShootExtensionStatusSpec
```
Comment (timuthy) [inline pkg/apis/core/types_shootextensionstatus.go]: WDYT about adding not using `Spec` but `ExtensionStatuses` directly here? I think it's somehow confusing to store a state (`ProviderStatus`) in a `Spec` field. In addition, there is no controller which would use this `Spec` and transfer it to an actual state, i.e., there is no `Spec <-> Status` pair. (/cc @rfranzke)

Comment (danielfoehrKn) [reply pkg/apis/core/types_shootextensionstatus.go]: Ok I see what you mean.
I modeled it after the ShootState resource as it is very similar mechanics.
So the argument to model it like this is only for consistency.
WDYT @timuthy

Comment (timuthy) [reply pkg/apis/core/types_shootextensionstatus.go]: Yes, I kind of wondered why we did it for the `ShootState` in the first place and maybe should change it there after some time.

I get the consistency argument but rather vote for following the K8s convention (see [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status)) if we now have the chance to do so:

> All objects that represent a physical resource whose state may vary from the user's desired intent SHOULD have a "spec" and a "status". Objects whose state cannot vary from the user's desired intent MAY have only "spec", and MAY rename "spec" to a more appropriate name.

Comment (rfranzke) [reply pkg/apis/core/types_shootextensionstatus.go]: I'd also vote for @timuthy's proposal. When we promote the `ShootState` to `v1beta1` we should do it similarly 👍🏻

Comment (danielfoehrKn) [reply pkg/apis/core/types_shootextensionstatus.go]: thanks for your input. Will adapt it

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Comment by rfranzke** [comment]
Needs another
/rebase
and some unit tests are failing, please check

/author-action

**Review by rfranzke — APPROVED** [review]
/lgtm

@timuthy - any final opinions?

**Review by timuthy — APPROVED** [review]
/lgtm
Thank you

**Pull request merged**